service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceOperations.java

```diff
@@ -141,9 +144,94 @@ public static HttpResponse request(
+     *
+     * @param iotHubConnectionString is the connection string for the IoTHub.
+     * @param url is the Twin URL for the device ID.
+     * @param method is the HTTP method (GET, POST, DELETE, PATCH, PUT).
+     * @param payload is the array of bytes that contains the payload.
+     * @param requestId is an unique number that identify the request.
+     * @param connectTimeout the http connect timeout to use.
+     * @param readTimeout the http read timeout to use.
+     * @param proxy the proxy to use, or null if no proxy will be used.
+     * @return the result of the request.
+     * @throws IotHubException This exception is thrown if the response verification failed.
+     * @throws IOException This exception is thrown if the IO operation failed.
+     */
+    public static HttpResponse request(
+            IotHubConnectionString iotHubConnectionString,
+            URL url,
+            HttpMethod method,
+            byte[] payload,
+            String requestId,
+            int connectTimeout,
+            int readTimeout,
+            Proxy proxy)
+            throws IOException, IotHubException, IllegalArgumentException
+    {
+        if (iotHubConnectionString == null)
+        {
+            throw new IllegalArgumentException("Null ConnectionString");
+        }
+
+        if (url == null)
+        {
+            throw new IllegalArgumentException("Null URL");
+        }
+
+        if (method == null)
+        {
+            throw new IllegalArgumentException("Null method");
+        }
+
+        String sasTokenString = new IotHubServiceSasToken(iotHubConnectionString).toString();
+        if((sasTokenString == null) || sasTokenString.isEmpty())
+        {
+            throw new IOException("Illegal sasToken null or empty");
+        }
+
+        HttpRequest request;
+        if (proxy != null)
+        {
+            request = new HttpRequest(url, method, payload, proxy);
+        }
+        else
+        {
+            request = new HttpRequest(url, method, payload);
+        }
+
+        request.setReadTimeoutMillis(readTimeout);
+        request.setConnectTimeoutMillis(connectTimeout);
+
+        if((requestId != null) && !requestId.isEmpty())
+        {
+            request.setHeaderField(REQUEST_ID, requestId);
+        }
+
+        request.setHeaderField(AUTHORIZATION, sasTokenString);
+        request.setHeaderField(USER_AGENT, TransportUtils.javaServiceClientIdentifier + TransportUtils.serviceVersion);
+        request.setHeaderField(ACCEPT, ACCEPT_VALUE);
+        request.setHeaderField(CONTENT_TYPE, ACCEPT_VALUE + "; " + ACCEPT_CHARSET);
+
+        if (headers != null)
+        {
+            for(Map.Entry<String, String> header : headers.entrySet())
+            {
+                request.setHeaderField(header.getKey(), header.getValue());
+            }
+
+            headers = null;
+        }
+
+        HttpResponse response = request.send();
```

Comment:
should we not catch exception here?

Reply:
No, what would we do with the exception if we caught it? For IOExceptions, they will get bubbled up to the user as expected. For hub level exceptions (404, 429, etc) the line below this one check the http response and will generate and throw any appropriate exceptions